Parser/Parser.cs

```diff
@@ -30,17 +30,33 @@ NBlock Block ()
 
    // declarations = [var-decls] [procfn-decls] .
    NDeclarations Declarations () {
-      List<NVarDecl> vars = new ();
-      if (Match (VAR)) 
+      List<NVarDecl> vars = new (); List<NFuncProcDecl> funcProcs = new ();
+      if (Match (VAR))
          do { vars.AddRange (VarDecls ()); Expect (SEMI); } while (Peek (IDENT));
-      return new (vars.ToArray ());
+
+      while (Match (FUNCTION) || Match (PROCEDURE)) {
+         var tok = mPrevious; var ident = Expect (IDENT); var parameters = ParamsList (); var type = None;
+         if (tok.Kind == FUNCTION) { Match (COLON); type = Type (); }
+         Expect (SEMI); var block = Block (); Expect (SEMI);
+         funcProcs.Add (new NFuncProcDecl (ident, parameters, type, block));
+      }
+      return new NDeclarations (vars.ToArray (), funcProcs.ToArray ());
    }
 
    // ident-list = IDENT { "," IDENT }
    Token[] IdentList () {
       List<Token> names = new ();
       do { names.Add (Expect (IDENT)); } while (Match (COMMA));
-      return names.ToArray (); 
+      return names.ToArray ();
+   }
+
+   // paramlist = "(" [var-decl { ";" var-decl }] ")"

```

Comment:
Unlike in a var declaration sequence, the variable groups in a parameter list are separated by commas:
`paramlist  =  "(" var-decl { "," var-decl } ")"`

Reply:
I checked with a online pascal compiler. Code doesn't compile when parameters are separated by comma. It expects semicolon. Found the same in few pascal documentations.

Reply:
OK, let's accept this as the standard definition.